From 44beaf40e6ab28cb50e1fad4778a1d4f17340f61 Mon Sep 17 00:00:00 2001
From: Suprabh Shukla <suprabh@google.com>
Date: Mon, 27 Jun 2022 14:01:16 -0700
Subject: [PATCH] Stop crashing the system on hitting the alarm limit

Exempting the system as a runtime restart is not clearly
better than extreme memory and computation pressure that can result from
the originating spam.
Callers in the system should guard against any spammy requests that lead
them to create a lot of alarms.

Test: Builds, boots and existing tests should pass.

atest CtsAlarmManagerTestCases:UidCapTests
atest FrameworksMockingServicesTests:AlarmManagerServiceTest

Bug: 234441463
Change-Id: Id5e94d44ac9ab24870a8213ec7583da0f592a5ff
(cherry picked from commit 3b9f3f4a0f5a661be65e287996cae8a4481a1453)
Merged-In: Id5e94d44ac9ab24870a8213ec7583da0f592a5ff
(cherry picked from commit c010da3a4649a02afe256cbf6dad475c2278059b)
(cherry picked from commit 909251a2caf1a8734c7e3a43794399ab3d15ac76)
Merged-In: Id5e94d44ac9ab24870a8213ec7583da0f592a5ff
---
 .../core/java/com/android/server/AlarmManagerService.java  | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/AlarmManagerService.java b/services/core/java/com/android/server/AlarmManagerService.java
index 4fab067d1d3b..b4ebd158c3b4 100644
--- a/services/core/java/com/android/server/AlarmManagerService.java
+++ b/services/core/java/com/android/server/AlarmManagerService.java
@@ -75,6 +75,7 @@ import android.text.format.DateFormat;
 import android.text.format.DateUtils;
 import android.util.ArrayMap;
 import android.util.ArraySet;
+import android.util.EventLog;
 import android.util.KeyValueListParser;
 import android.util.Log;
 import android.util.LongArrayQueue;
@@ -1776,7 +1777,11 @@ class AlarmManagerService extends SystemService {
                 if (DEBUG_PER_UID_LIMIT && UserHandle.isCore(callingUid)) {
                     logAllAlarmsForUidLocked(callingUid);
                 }
-                throw new IllegalStateException(errorMsg);
+                if (callingUid != Process.SYSTEM_UID) {
+                    throw new IllegalStateException(errorMsg);
+                } else {
+                    EventLog.writeEvent(0x534e4554, "234441463", -1, errorMsg);
+                }
             }
             setImplLocked(type, triggerAtTime, triggerElapsed, windowLength, maxElapsed,
                     interval, operation, directReceiver, listenerTag, flags, true, workSource,
-- 
2.38.0.413.g74048e4d9e-goog

